src/app/src/test/kotlin/org/vaccineimpact/orderlyweb/tests/unit_tests/templates/WebLoginTest.kt

```diff
@@ -0,0 +1,53 @@
+package org.vaccineimpact.orderlyweb.tests.unit_tests.templates
+
+import com.nhaarman.mockito_kotlin.mock
+import org.assertj.core.api.Assertions
+import org.junit.ClassRule
+import org.junit.Test
+import org.vaccineimpact.orderlyweb.ActionContext
+import org.vaccineimpact.orderlyweb.db.AppConfig
+import org.vaccineimpact.orderlyweb.tests.unit_tests.templates.rules.FreemarkerTestRule
+import org.vaccineimpact.orderlyweb.viewmodels.WebloginViewModel
+
+class WebLoginTest
+{
+    companion object
+    {
+        @ClassRule
+        @JvmField
+        val template = FreemarkerTestRule("weblogin.ftl")
+    }
+
+    private val testModel = WebloginViewModel(mock<ActionContext>(), "/fakepath")
+    private val doc = template.jsoupDocFor(testModel)
+
+    @Test
+    fun `renders page`()
+    {
+        Assertions.assertThat(doc.select("#external-login").count()).isEqualTo(1)
+    }
+
+    @Test
+    fun `renders breadcrumbs correctly`()
+    {
+        val breadcrumbs = doc.select(".crumb-item")
+
+        Assertions.assertThat(breadcrumbs.count()).isEqualTo(2)
+        Assertions.assertThat(breadcrumbs[0].child(0).text()).isEqualTo("Main menu")
+        Assertions.assertThat(breadcrumbs[0].child(0).attr("href")).isEqualTo("http://localhost:8888")
+        Assertions.assertThat(breadcrumbs[1].child(0).text()).isEqualTo("Login")
+        Assertions.assertThat(breadcrumbs[1].child(0).attr("href")).isEqualTo("/weblogin")
+    }
+
+    @Test
+    fun `renders external link correctly`()
+    {
+        val link = doc.select(".login-link")
+
+        Assertions.assertThat(link.count()).isEqualTo(1)
+        Assertions.assertThat(link.attr("href"))
+            .isEqualTo("http://localhost:8888/weblogin/external?requestedUrl=/fakepath")
+
+        Assertions.assertThat(link.text()).isEqualToIgnoringCase("Log in with ${AppConfig()["auth.provider"]}")
```

Comment:
Ingenious way of admitting where this 'unit test' needs to get its config from! :) 
Could use String.capitalize() rather than ignoring the case..

Reply:
expected text is "GitHub" not Github so string capitalise won't work.

Reply:
Ahh, so it is